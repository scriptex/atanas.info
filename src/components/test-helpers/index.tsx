import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '../../scripts/shared';

export const mockUseEffect = (): void => {
	let useEffect: jest.Mock<React.EffectCallback>;

	const mockedUseEffect = () => {
		useEffect.mockImplementationOnce((f: () => React.EffectCallback) => f());
	};

	beforeEach(() => {
		useEffect = jest.spyOn(React, 'useEffect') as unknown as jest.Mock<React.EffectCallback>;

		mockedUseEffect();
	});
};

export const snapshotTest = (
	Component: React.FC<Readonly<React.ComponentProps<any>>>,
	buttonSelector?: string
): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	mockUseEffect();

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { asFragment, container } = result;

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				act(() => {
					container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { unmount, asFragment } = result;

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};
