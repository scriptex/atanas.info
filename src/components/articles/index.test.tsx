import { Articles } from '.';
import { snapshotTest } from '../test-helpers';

snapshotTest(Articles);
