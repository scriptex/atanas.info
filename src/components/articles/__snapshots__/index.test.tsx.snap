// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles component Should render the Articles component 1`] = `
<DocumentFragment>
  <main
    class="o-main"
  >
    <section
      class="c-section c-section--articles c-section--slides fullsize-background"
      id="articles"
      style="background-image: url(images/temp/articles.jpg);"
    >
      <div
        class="o-shell"
      >
        <header
          class="c-section__header"
        >
          <div
            class="c-section__animation"
            style="width: 150px; height: 150px; margin-left: auto; margin-right: auto;"
          />
          <h2>
            Articles
          </h2>
        </header>
        <div
          class="c-section__body o-grid"
        >
          <div
            class="o-grid__item xs-12 sm-6"
          >
            <a
              class="c-article-link fullsize-background"
              href="https://htmlburger.com/blog/jarvis-task-runner/"
              rel="noopener noreferrer"
              style="background-image: url(https://htmlburger.com/blog/wp-content/uploads/2017/11/jarvis-Recovered-2-826x464.jpg);"
              target="_blank"
            >
              <strong>
                Jarvis: The Intelligent Front-End Task Runner
              </strong>
              <svg
                class="c-svg-external-link"
              >
                <use
                  xlink:href="#svg-external-link"
                />
              </svg>
            </a>
            <div
              class="c-loading"
            >
              <div
                class="c-loader"
              />
            </div>
          </div>
          <div
            class="o-grid__item xs-12 sm-6"
          >
            <a
              class="c-article-link fullsize-background"
              href="https://htmlboutique.com/blog/2015/11/13/ionic-development-by-htmlboutique/"
              rel="noopener noreferrer"
              style="background-image: url(https://htmlboutique.com/blog/wp-content/uploads/2015/11/ionicnew3.jpg);"
              target="_blank"
            >
              <strong>
                Ionic Development by htmlBoutique
              </strong>
              <svg
                class="c-svg-external-link"
              >
                <use
                  xlink:href="#svg-external-link"
                />
              </svg>
            </a>
            <div
              class="c-loading"
            >
              <div
                class="c-loader"
              />
            </div>
          </div>
        </div>
        <div
          class="c-section__actions"
        >
          <button
            class="c-btn"
            type="button"
          >
            Get in touch
          </button>
          <ul
            class="c-section__actions-list"
          >
            <li>
              <button
                class="c-btn c-section__actions-close"
                type="button"
              >
                Close
              </button>
            </li>
            <li>
              <button
                class="c-btn"
                type="button"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-email"
                  />
                </svg>
                <span>
                  Contact me
                </span>
                <strong>
                  directly
                </strong>
              </button>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://app.usebraintrust.com/talent/782/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-braintrust"
                  />
                </svg>
                <span>
                  Hire me on
                </span>
                <strong>
                  Braintrust
                </strong>
                <svg
                  class="c-svg-external-link"
                >
                  <use
                    xlink:href="#svg-external-link"
                  />
                </svg>
              </a>
            </li>
            <li>
              <a
                class="c-btn"
                href="https://www.toptal.com/resume/atanas-atanasov"
                rel="noopener noreferrer"
                target="_blank"
              >
                <svg
                  class="c-section__actions-icon"
                >
                  <use
                    xlink:href="#svg-toptal"
                  />
                </svg>
                <span>
                  Hire me on
                </span>
                <strong>
                  Toptal
                </strong>
                <svg
                  class="c-svg-external-link"
                >
                  <use
                    xlink:href="#svg-external-link"
                  />
                </svg>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </section>
  </main>
</DocumentFragment>
`;

exports[`Articles component Should unmount the Articles component 1`] = `<DocumentFragment />`;
